pkg/applyconfigurations/gen.go

```diff
@@ -0,0 +1,230 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package applyconfigurations
+
+import (
+	"fmt"
+	"go/ast"
+	"os"
+	"path/filepath"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/code-generator/cmd/applyconfiguration-gen/args"
+	"k8s.io/code-generator/cmd/applyconfiguration-gen/generators"
+
+	"k8s.io/gengo/v2"
+	"k8s.io/gengo/v2/generator"
+	"k8s.io/gengo/v2/parser"
+
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/loader"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+// Based on deepcopy gen but with legacy marker support removed.
+
+var (
+	isCRDMarker      = markers.Must(markers.MakeDefinition("kubebuilder:resource", markers.DescribesType, crdmarkers.Resource{}))
+	enablePkgMarker  = markers.Must(markers.MakeDefinition("kubebuilder:ac:generate", markers.DescribesPackage, false))
+	outputPkgMarker  = markers.Must(markers.MakeDefinition("kubebuilder:ac:output:package", markers.DescribesPackage, ""))
+	enableTypeMarker = markers.Must(markers.MakeDefinition("kubebuilder:ac:generate", markers.DescribesType, false))
+)
+
+const importPathSuffix = "applyconfiguration"
+
+// +controllertools:marker:generateHelp
+
+// Generator generates code containing apply configuration type implementations.
+type Generator struct {
+	// HeaderFile specifies the header text (e.g. license) to prepend to generated files.
+	HeaderFile string `marker:",optional"`
+}
+
+func (Generator) CheckFilter() loader.NodeFilter {
+	return func(node ast.Node) bool {
+		// ignore interfaces
+		_, isIface := node.(*ast.InterfaceType)
+		return !isIface
+	}
+}
+
+func (Generator) RegisterMarkers(into *markers.Registry) error {
+	if err := markers.RegisterAll(into,
+		isCRDMarker, enablePkgMarker, enableTypeMarker, outputPkgMarker); err != nil {
+		return err
+	}
+
+	into.AddHelp(isCRDMarker,
+		markers.SimpleHelp("apply", "enables apply configuration generation for this type"))
+	into.AddHelp(
+		enableTypeMarker, markers.SimpleHelp("apply", "overrides enabling or disabling applyconfigurations generation for the type"))
+	into.AddHelp(
+		enablePkgMarker, markers.SimpleHelp("apply", "overrides enabling or disabling applyconfigurations generation for the package"))
+	into.AddHelp(
+		outputPkgMarker, markers.SimpleHelp("apply", "overrides the default output package for the applyconfigurations generation"))
+	return nil
+}
+
+func enabledOnPackage(col *markers.Collector, pkg *loader.Package) (bool, error) {
+	pkgMarkers, err := markers.PackageMarkers(col, pkg)
+	if err != nil {
+		return false, err
+	}
+	pkgMarker := pkgMarkers.Get(enablePkgMarker.Name)
+	if pkgMarker != nil {
+		return pkgMarker.(bool), nil
+	}
+	return false, nil
+}
+
+// enableOnType marks whether applyconfiguration generation is enabled for the type.
```

Comment:
nit: "marks" might be the wrong verb

(I'm also fine with no godoc comment here, the func name is good enough)

Same in l.115